include/fastgltf/types.hpp

```diff
@@ -1170,31 +1183,67 @@ namespace fastgltf {
 		}
 
 		template <typename F>
-		[[nodiscard]] OptionalWithFlagValue<T> and_then(F&& func)& {
+		[[nodiscard]] auto and_then(F&& func)& {
+			using U = std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<F, T&>>>;
+			if (!has_value())
+				return U();
+			return std::invoke(std::forward<F>(func), **this);
+		}
+
+		template <typename F>
+		[[nodiscard]] auto and_then(F&& func) const& {
+			using U = std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<F, const T&>>>;
+			if (!has_value())
+				return U();
+			return std::invoke(std::forward<F>(func), **this);
+		}
+
+		template <typename F>
+		[[nodiscard]] auto and_then(F&& func)&& {
+			using U = std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<F, T&&>>>;
+			if (!has_value())
+				return U();
+			return std::invoke(std::forward<F>(func), std::move(**this));
+		}
+
+		template <typename F>
+		[[nodiscard]] auto and_then(F&& func) const&& {
+			using U = std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<F, const T&&>>>;
+			if (!has_value())
+				return U();
+			return std::invoke(std::forward<F>(func), std::move(**this));
+		}
+
+		template <typename F>
+		[[nodiscard]] auto transform(F&& func)& {
+			using U = std::remove_cv_t<std::invoke_result_t<F, T&>>;
 			if (!has_value())
-				return std::nullopt;
-			return func(value());
+				return Optional<U>();
+			return Optional<U>(std::invoke(std::forward<F>(func), **this));
 		}
 
 		template <typename F>
-		[[nodiscard]] OptionalWithFlagValue<T> and_then(F&& func)&& {
+		[[nodiscard]] auto transform(F&& func) const& {
+			using U = std::remove_cv_t<std::invoke_result_t<F, const T&>>;
 			if (!has_value())
-				return std::nullopt;
-			return func(std::move(value()));
+				return Optional<U>();
+			return Optional<U>(std::invoke(std::forward<F>(func), **this));
 		}
 
 		template <typename F>
-		[[nodiscard]] OptionalWithFlagValue<T> transform(F&& func)& {
+		[[nodiscard]] auto transform(F&& func)&& {
+			using U = std::remove_cv_t<std::invoke_result_t<F, T&&>>;
 			if (!has_value())
-				return std::nullopt;
-			return OptionalWithFlagValue<T>(func(value()));
+				return Optional<U>();
+			return Optional<U>(std::invoke(std::forward<F>(func), std::move(**this)));
 		}
 
 		template <typename F>
-		[[nodiscard]] OptionalWithFlagValue<T> transform(F&& func)&& {
+		[[nodiscard]] auto transform(F&& func) const&& {
+			using U = std::remove_cv_t<std::invoke_result_t<F, const T&&>>;
 			if (!has_value())
-				return std::nullopt;
-			return OptionalWithFlagValue<T>(func(std::move(value())));
+				return Optional<U>();
+			return Optional<U>(std::invoke(std::forward<F>(func), std::move(**this)));
```

Comment:
I think the values for U here are wrong, at least if I look at https://en.cppreference.com/w/cpp/utility/optional/transform.

For the `&&` function it should just be `T`, and for the `const&&` function it should be `const T`?

Reply:
Oh, yes. I was looking at Clang's implementation, so maybe they messed up as well 😂